### PR TITLE
Email change confirmation For Casa Admins changing User's email

### LINKED ISSUE #4781: Add confirmation process for Casa Admins changing Users Emails
**What type(s) of user does this feature affect?**
 - volunteers? Yes
 - supervisors? Yes
 - admins? YES
 - all casa admins? YES

**Description**
Part 2 of issue #4594
As part of the User Export Report/User Email Change update, the email confirmation flow needs to be applied to Casa Admins/AllCasaAdmins changing the emails of other users. Currently, there is no confirmation process in place and thus needs to brought up to industry standards.

**Screenshots of current behavior, if any**  
You can paste images on the clipboard here
Screenshot 2023-04-25 at 9 37 10 AM
Screenshot 2023-04-25 at 9 37 23 AM

_Login Details:_  
[Link to QA site](https://casa-qa.herokuapp.com/)  

Login Emails: 
- volunteer1@example.com  view site as a volunteer
- supervisor1@example.com view site as a supervisor
- casa_admin1@example.com view site as an admin
- all_casa_admin1@example.com view site as an all casa admin
  - go to `/all_casa_admins/sign_in`  

password for all users: 12345678  

### Questions? Join Slack!

We highly recommend that you join us in slack https://rubyforgood.herokuapp.com/ #casa channel to ask questions quickly and hear about office hours (currently Tuesday 5-7pm Pacific), stakeholder news, and upcoming new issues.

### PR DESCRIPTION
**Pull request by ShamiTomita**
### What github issue is this PR for, if any?
Resolves #4781 

### What changed, and why?
- [X] Removed the skip_confirmation! method from the update method for volunteers, supervisors, casa_admins.
- [X] Confirmation Emails are sent to the relevant user when a Casa Admin changes that user's email
- [X] Included two ApplicationController methods to check and properly notify the current_user after updating another user's profile. 
- [X] Updated relevant request and system edit specs to check for confirmation emails after email change submissions and then separately checking for successful email change upon confirmation.

A Future PR will most likely involve applying the same confirmation process to All Casa Admins. This would require adding Confirmable to All Casa Admins.

### How will this affect user permissions?
- Volunteer permissions:
- Supervisor permissions:
- Admin permissions: Can no longer completely change other user's emails. They can only being the email change process. 

### How is this tested? (please write tests!) 💖💪
Updates to relevant system and request specs for Volunteers, Supervisors, Admins, (And ALL_CASA_ADMINS changing other CASA Admins' emails)

### Screenshots please :)
Application Controller Methods (so all user role controllers have access)
Screenshot 2023-05-09 at 6 38 41 PM

Where these changes are applied:
Screenshot 2023-04-28 at 2 56 11 PM
Screenshot 2023-04-28 at 2 57 11 PM
Screenshot 2023-04-28 at 2 57 29 PM
Screenshot 2023-04-28 at 2 57 42 PM

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`alt text`

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

### CODE REVIEW

**Comment by ShamiTomita** [comment]
@cnorm35 I think this is ready for a review! LMK if there are any change I should making before submitting for final review.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -107,4 +107,16 @@ def not_authorized
     flash[:notice] = "Sorry, you are not authorized to perform this action."
     redirect_to(root_url)
   end
+
+  def admin_email_change(user) # for casa admins
+    if user.saved_changes.include?("unconfirmed_email")
+      redirect_to edit_polymorphic_path(user), notice: "Confirmation Email Sent To #{user.role}."
```
Comment (cnorm35) [inline app/controllers/application_controller.rb:113]: 🏅  love this, great way to keep this clean

File: app/controllers/volunteers_controller.rb [hunk 2]
```diff
@@ -51,10 +51,11 @@ def edit
 
   def update
     authorize @volunteer
-    @volunteer.skip_reconfirmation!
     if @volunteer.update(update_volunteer_params)
-      @volunteer.filter_old_emails!(@volunteer.email)
-      redirect_to edit_volunteer_path(@volunteer), notice: "Volunteer was successfully updated."
+      unless admin_email_change(@volunteer) # this method is in application controller
```
Comment (cnorm35) [inline app/controllers/volunteers_controller.rb:55]: I like the idea of the comment but think we can skip the comment.  If this method was defined somewhere _besides_ the `ApplicationController` I would say add a comment but think this is readable enough without it.

File: spec/system/casa_admins/edit_spec.rb [hunk 3]
```diff
@@ -6,27 +6,53 @@
...
       fill_in "Display name", with: expected_display_name
       fill_in "Phone number", with: expected_phone_number
 
       click_on "Submit"
 
       admin.reload
 
-      expect(page).to have_text "New admin created successfully"
-      expect(admin.email).to eq expected_email
+      expect(page).to have_text "New admin created successfully" # should be updated
```
Comment (cnorm35) [inline spec/system/casa_admins/edit_spec.rb:22]: stray comment here

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (ShamiTomita) [reply app/controllers/application_controller.rb:113]: Yay! Thank you! Big thanks to Code Climate for prodding me to go for it

File: app/controllers/volunteers_controller.rb (hunk 2 above)
Comment (ShamiTomita) [reply app/controllers/volunteers_controller.rb:55]: Removed!

File: spec/system/casa_admins/edit_spec.rb (hunk 3 above)
Comment (ShamiTomita) [reply spec/system/casa_admins/edit_spec.rb:22]: Removed!

File: spec/system/casa_admins/edit_spec.rb [hunk 4]
```diff
@@ -6,27 +6,53 @@
...
+  context "with valid email data" do
+    before do
+      visit edit_casa_admin_path(admin)
+      @old_email = admin.email
+      fill_in "Email", with: "new_admin_email@example.com"
+
+      click_on "Submit"
+      admin.reload
+    end
+    it "sends a confirmation email upon submission and does not change the user's displayed email" do
```
Comment (cnorm35) [inline spec/system/casa_admins/edit_spec.rb:37]: it looks like most of these other system specs have a space between each expectation and think that makes it more readable.

need to add spaces between the `end` of the previous spec and the `it` for the current one.

File: spec/system/supervisors/edit_spec.rb [hunk 5]
```diff
@@ -156,18 +156,34 @@
...
+        @supervisor = create(:supervisor)
+        @old_email = @supervisor.email
+        visit edit_supervisor_path(@supervisor)
+        fill_in "supervisor_email", with: ""
+        fill_in "supervisor_email", with: "new_supervisor_email@example.com"
+
+        click_on "Submit"
+        @supervisor.reload
+      end
+      it "sends a confirmaton email to the supervisor and displays current email" do
```
Comment (cnorm35) [inline spec/system/supervisors/edit_spec.rb:170]: add space between expectations

File: spec/system/volunteers/edit_spec.rb [hunk 6]
```diff
@@ -54,6 +53,39 @@
...
+        expect(ActionMailer::Base.deliveries.count).to eq(1)
+        expect(ActionMailer::Base.deliveries.first).to be_a(Mail::Message)
+        expect(ActionMailer::Base.deliveries.first.body.encoded)
+          .to match("You can confirm your account email through the link below:")
+
+        expect(page).to have_text "Confirmation Email Sent To Volunteer."
+        expect(page).to have_field("Email", with: @old_email)
+        expect(volunteer.unconfirmed_email).to eq("newemail@example.com")
+      end
+      it "succesfully displays the new email once the user confirms" do
```
Comment (cnorm35) [inline spec/system/volunteers/edit_spec.rb:77]: add space between expectations

**Review by cnorm35 — CHANGES_REQUESTED** [review]
This looks great @ShamiTomita 

I added a few change requests to help with the readability but everything else is good to go.

Edit: once you've added all your changes, don't forget to remove WIP from the pull request title so everyone else can know this is ready for review.

File: spec/system/casa_admins/edit_spec.rb (hunk 4 above)
Comment (ShamiTomita) [reply spec/system/casa_admins/edit_spec.rb:38]: Totally agree, for some reason I thought one of the linters would help with that. Added the spaces.

File: spec/system/supervisors/edit_spec.rb (hunk 5 above)
Comment (ShamiTomita) [reply spec/system/supervisors/edit_spec.rb:171]: added

File: spec/system/volunteers/edit_spec.rb (hunk 6 above)
Comment (ShamiTomita) [reply spec/system/volunteers/edit_spec.rb:78]: added

**Comment by ShamiTomita** [comment]
Thank you so much @cnorm35! I'll make sure to update the title.

**Review by ShamiTomita — COMMENTED** [review]
Good To Go on my end!

**Review by cnorm35 — APPROVED** [review]

**Comment by FireLemons** [comment]
:1st_place_medal:

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -107,4 +107,20 @@ def not_authorized
...
+    notice
+  end
+
+  def check_all_casa_admin_email_change_confirmation_notice(user) # for all casa admins
+    notice = "#{user.role} was successfully updated."
+    if user.saved_changes.include?("unconfirmed_email")
+      notice += " Confirmation Email Sent."
+    end
+    notice
+  end
```
Comment (littleforest) [inline app/controllers/application_controller.rb:125]: Adding my same comment as the other PR: These methods look exactly the same to me, are they not? Can we get rid of one and maybe make the name of the method more generic?

Comment (ShamiTomita) [reply app/controllers/application_controller.rb:117]: Applied the changes!

**Review by littleforest — APPROVED** [review]
👍 🎉

**Pull request merged**